Tools/scripts/summarize_stats.py

```diff
@@ -2,908 +2,988 @@
 default stats folders.
 """
 
+from __future__ import annotations
+
 # NOTE: Bytecode introspection modules (opcode, dis, etc.) should only
-# happen when loading a single dataset. When comparing datasets, it
+# be imported when loading a single dataset. When comparing datasets, it
 # could get it wrong, leading to subtle errors.
 
 import argparse
 import collections
-import json
-import os.path
 from datetime import date
 import itertools
-import sys
+import json
+import os
+from pathlib import Path
 import re
+import sys
+from typing import TextIO, TypeAlias
+
+
+OpcodeStats: TypeAlias = dict[str, dict[str, int]]
+PairCounts: TypeAlias = list[tuple[int, tuple[str, str]]]
+Defines: TypeAlias = dict[int, list[str]]
+Rows: TypeAlias = list[tuple]
+Columns: TypeAlias = tuple[str, ...]
+
 
 if os.name == "nt":
     DEFAULT_DIR = "c:\\temp\\py_stats\\"
 else:
     DEFAULT_DIR = "/tmp/py_stats/"
 
-TOTAL = "specialization.hit", "specialization.miss", "execution_count"
 
+SOURCE_DIR = Path(__file__).parents[2]
 
-def format_ratio(num, den):
-    """
-    Format a ratio as a percentage. When the denominator is 0, returns the empty
-    string.
-    """
-    if den == 0:
-        return ""
-    else:
-        return f"{num/den:.01%}"
 
+TOTAL = "specialization.hit", "specialization.miss", "execution_count"
 
-def percentage_to_float(s):
-    """
-    Converts a percentage string to a float.  The empty string is returned as 0.0
-    """
-    if s == "":
-        return 0.0
-    else:
-        assert s[-1] == "%"
-        return float(s[:-1])
 
+def pretty(name: str) -> str:
+    return name.replace("_", " ").lower()
 
-def join_rows(a_rows, b_rows):
-    """
-    Joins two tables together, side-by-side, where the first column in each is a
-    common key.
-    """
-    if len(a_rows) == 0 and len(b_rows) == 0:
-        return []
 
-    if len(a_rows):
-        a_ncols = list(set(len(x) for x in a_rows))
-        if len(a_ncols) != 1:
-            raise ValueError("Table a is ragged")
+class Stats(dict):
+    def __init__(self, input: Path):
+        super().__init__()
 
-    if len(b_rows):
-        b_ncols = list(set(len(x) for x in b_rows))
-        if len(b_ncols) != 1:
-            raise ValueError("Table b is ragged")
+        if input.is_file():
+            with open(input, "r") as fd:
+                self.update(json.load(fd))
 
-    if len(a_rows) and len(b_rows) and a_ncols[0] != b_ncols[0]:
-        raise ValueError("Tables have different widths")
+            self["_stats_defines"] = {
+                int(k): v for k, v in self["_stats_defines"].items()
+            }
+            self["_defines"] = {int(k): v for k, v in self["_defines"].items()}
 
-    if len(a_rows):
-        ncols = a_ncols[0]
-    else:
-        ncols = b_ncols[0]
+        elif input.is_dir():
+            stats: collections.Counter = collections.Counter()
```

Comment:
This type annotation seems redundant

Reply:
Yep, but mypy requires it :(

Reply:
Don't use mypy then?

Is there a Mypy issue for this?

Reply:
This should make mypy happy and Mark less unhappy:

```suggestion
            stats = collections.Counter[str]()
```

Reply:
> Don't use mypy then?
> 
> Is there a Mypy issue for this?

It's not a mypy bug. Mypy can't tell what kind of items are going to be stored as keys for the `Counter`, so it demands an explicit annotation. @mdboom's annotation shuts mypy up, because `collections.Counter` as a type annotation is equivalent to `collections.Counter[Any]`. But the better solution is to use `collections.Counter[str]`, because they keys of `stats` are all strings.

Reply:
It is a mypy bug.
Omitting the annotation and providing `collections.Counter` as an annotation has exactly the same information content. So complaining about one and not the other is erroneous.
I assume mypy will not complain about `stats = collections.Counter[str]()` then.

Reply:
I agree there's a usability bug here; mypy isn't communicating what it wants from you very clearly at all.

> I assume mypy will not complain about `stats = collections.Counter[str]()` then.

Correct, that will make mypy happy (https://github.com/python/cpython/pull/110398#discussion_r1347626014)